Add ability to auto scale based on residuals

lindsayad opened the pull request:
Needs some tests...

Closes #14397

lindsayad committed:
Add ability to auto scale based on residuals

Closes #14397

lindsayad commented:
@bwspenc @novasr @dschwen We should try this out on the BISON assessment...

novasr commented:
@elementx54, try running the assessment case you showed us this week, as committed, but with this branch.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/435046/) on 2d614b9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14422/site)

This comment will be updated on new commits.

lindsayad commented:
This is of course why tests are so valuable 😄 In the process of adding a test, I discovered that I definitely gummed up the scaling in parallel...fix coming soon...

lindsayad committed:
Add single variable test

lindsayad committed:
MooseVec(Mat)View

Keep declarations in Moose.h so that they are visible in any
translation unit that we are debugging. However, move definitions
to PetscSupport because that's the place where @fdkong likes
to have PETSc code

lindsayad committed:
Make sure we don't compute NodalBCs

lindsayad committed:
Add two-var tests

permcody commented on framework/include/base/Moose.h:
```diff
@@ -32,8 +32,12 @@ class MooseEnumItem;
 class ExecFlagEnum;
 class MooseVariableFEBase;
 
+#ifdef LIBMESH_HAVE_PETSC
```
Can we just add this to MOOSE already?
```
#ifndef LIBMESH_HAVE_PETSC
   YOUR DUMB
#endif 
```

lindsayad replied:
I agree...

lindsayad replied:
I agree...

permcody replied:
Wow, it appears you doubly agree! Is that message PC enough?

lindsayad replied:
Stupid github slowness. Wow this is the most constructive review I've ever had 😆

lindsayad commented:
People asked for this; it’s weird that no one has reviewed it yet 

> On Dec 25, 2019, at 1:23 PM, stale[bot] <notifications@github.com> wrote:
> 
> ﻿
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bwspenc commented:
I assume nobody is looking at it because this may prove to not be necessary with the fixes that you are working on for contact with scaling.

fdkong commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -653,6 +653,16 @@ class NonlinearSystemBase : public SystemBase,
+  {
+    _resid_vs_jac_scaling_param = resid_vs_jac_scaling_param;
+  }
+
```
0.5 is a perfect combination of both?

lindsayad replied:
Oh yes perfect

fdkong commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -462,15 +495,52 @@ NonlinearSystem::computeScalingJacobian()
+      else
+        inverse_scaling_factors[i] =
+            std::exp(_resid_vs_jac_scaling_param * std::log(resid_inverse_scaling_factors[i]) +
+                     (1 - _resid_vs_jac_scaling_param) * std::log(jac_inverse_scaling_factors[i]));
```
`inverse_scaling_factors[i] = _resid_vs_jac_scaling_param *resid_inverse_scaling_factors[i] + (1-_resid_vs_jac_scaling_param)*jac_inverse_scaling_factors[i]`

won't work?

lindsayad replied:
No because if say the Jacobian recommends scaling by 1e10 and residual recommends scaling by 1, we would then scale by 5e9 which is hardly a compromise. (This is assuming a choice of 0.5 for the `_resid_vs_jac_scaling_param`)

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1456,6 +1460,12 @@ NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)
+    // any strongly enforced constraint conditions or penalty-type objects
+    // (DGKernels, IntegratedBCs, InterfaceKernels, Constraints)
+    return;
+
```
DG simply won't work with scaling?

lindsayad replied:
DG can sometimes use penalty factors which is bad for determining a proper scaling factor. Moreover, except if a user is doing DG with pure advection (no diffusion, no sources, no sinks) and constant monomial basis functions, they are going to have regular Kernels in addition to their DGKernels. This is because there are elemental contributions in addition to the face contributions.

fdkong commented:
@lindsayad Looks great for me. Could you rebase?

lindsayad commented:
Ok, for an input based on @acasagran's multiphysics input from https://github.com/idaholab/moose/pull/14546#issuecomment-573960767, residual based automatic scaling works the best. The simulation runs to completion with `resid_vs_jac_scaling_param=1`. For pure Jacobian scaling, the simulation hits dtmin at time step 8. With hybrid scaling, `resid_vs_jac_scaling_param=0.5`, the simulation hits dtmin at time step 23.

It's clear that modules people setup simulations with parameters that are tuned for optimal convergence because I should note that without any scaling at all, the simulation solves in a total of 38 non-linear iterations. With automatic pure residual scaling, this simulation takes 43 non-linear iterations.

Anyway I think that this example demonstrates that pure residual automatic scaling is worthwhile.

lindsayad commented:
> It's clear that modules people setup simulations with parameters that are tuned for optimal convergence because I should note that without any scaling at all, the simulation solves in a total of 38 non-linear iterations. With automatic pure residual scaling, this simulation takes 43 non-linear iterations.

So if I change the `penalty` to match `youngs_modulus`, which would be more consistent with the `penalty` docstring: `The penalty to apply.  This can vary depending on the stiffness of your materials`, then the no scaling case takes 45 nonlinear its while the residual-based auto scaling takes 37 nonlinear iterations. Nice job auto scaling!

permcody commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -168,7 +168,7 @@ NonlinearSystemBase::NonlinearSystemBase(FEProblemBase & fe_problem,
     _compute_dampers_timer(registerTimedSection("computeDampers", 3)),
     _compute_dirac_timer(registerTimedSection("computeDirac", 3)),
-    _compute_scaling_jacobian_timer(registerTimedSection("computeScalingJacobian", 2)),
+    _compute_scaling_timer(registerTimedSection("computeScaling", 2)),
```
Make sure you initialize your new class member Boolean here as well just in case we refactor and it doesn't get set later.

lindsayad replied:
Do you mean the new `Real _resid_vs_scaling_param` class member. I don't think I added any new uninitialized booleans...

permcody reviewed (commented):
This looks fine with the exception of the one missing Boolean initialization to me assuming @fdkong approves. I can confirm the the Mammoth failure is unrelated and should be fixed shortly.

moosebuild commented:
Job [App tests PETSc alt](https://civet.inl.gov/job/452686/) on be91285 : invalidated by @fdkong

Marmmoth diffs

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/453806/) on f0ab7f0 : invalidated by @lindsayad

lindsayad commented:
@fdkong @permcody I think this should be set to go

permcody commented:
Roger... Just need @aeslaughter to fix next so we don't have a bunch of cascading failed builds.

permcody commented:
Ugh - It's amazing how crippling it is to only have ~20 build boxes up and running...

permcody commented:
@lindsayad - It looks like some of the new tests are failing in the sweeps. Maybe your scaling factors are changing a bit?

lindsayad commented:
Yea it looks that way. I'll look into it now

On Thu, Jan 23, 2020 at 10:45 AM Cody Permann <notifications@github.com>
wrote:

> @lindsayad <https://github.com/lindsayad> - It looks like some of the new
> tests are failing in the sweeps. Maybe your scaling factors are changing a
> bit?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14422?email_source=notifications&email_token=ACOGA4AW7USMIAPOYYOON5LQ7HQUVA5CNFSM4JQJAMN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJYNA3I#issuecomment-577818733>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4EWNNJHTMNHHPQ2B2DQ7HQUVANCNFSM4JQJAMNQ>
> .
>
